src/persist-client/src/cfg.rs

```diff
@@ -596,3 +596,41 @@ impl RustType<ConfigUpdates> for PersistParameters {
         Ok(Self { config_updates })
     }
 }
+
+// If persist gets some encoded ProtoState from the future (e.g. two versions of
+// code are running simultaneously against the same shard), it might have a
+// field that the current code doesn't know about. This would be silently
+// discarded at proto decode time. Unknown Fields [1] are a tool we can use in
+// the future to help deal with this, but in the short-term, it's best to keep
+// the persist read-modify-CaS loop simple for as long as we can get away with
+// it (i.e. until we have to offer the ability to do rollbacks).
+//
+// [1]: https://developers.google.com/protocol-buffers/docs/proto3#unknowns
+//
+// To detect the bad situation and disallow it, we tag every version of state
+// written to consensus with the version of code used to encode it. Then at
+// decode time, we're able to compare the current version against any we receive
+// and assert as necessary.
+//
+// Initially we allow any from the past (permanent backward compatibility) and
+// one minor version into the future (forward compatibility). This allows us to
+// run two versions concurrently for rolling upgrades. We'll have to revisit
+// this logic if/when we start using major versions other than 0.
+//
+// We could do the same for blob data, but it shouldn't be necessary. Any blob
+// data we read is going to be because we fetched it using a pointer stored in
+// some persist state. If we can handle the state, we can handle the blobs it
+// references, too.
+pub fn is_data_version_invalid(code_version: &Version, data_version: &Version) -> bool {
```

Comment:
nit: this isn't checking if a data version is valid as much as if this version of code can understand data written by processes built at that version. I'm having a hard time coming up with an alternative to propose besides just using `check_data_version` for both 😅. I'd also make the return type a `Result<(), String>` because it's too easy to get the meaning of the bool inverted

Reply:
Done.